[20138] use absolute urls for header linking in ToCs

floriank opened the pull request:
This will use the absolute URLs to a resource for rendering the anchors. It
prevents the interfering of the Angular router with the Url, leading to
faulty redirects.

It also removes the links when no request is present (in the case of mailers).

Should meet the requirements of https://community.openproject.org/work_packages/20138

Commit:
use absolute urls for header linking in ToCs

This will use the absolute URLs to a resource for rendering the anchors. It
prevents the interfering of the Angular router with the Url, leading to
faulty redirects.

It also removes the links when no request is present (in the case of mailers)

Signed-off-by: Florian Kraft <f.kraft@finn.de>

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://floriank-fix-20138-toc-links-broken.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://floriank-fix-20138-toc-links-broken.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

ulferts commented:
Hm, travis doesn't like it...at all.

floriank commented:
seems like a rebase issue..

NobodysNightmare commented on lib/open_project/text_formatting.rb:
```diff
@@ -418,5 +425,19 @@ def replace_toc(text, headings)
+    # displays the current url plus an optional anchor
+    #
+    def full_url(anchor_name = '')
+      return false if current_request.nil?
```
```
> false.to_s
 => "false" 
```

I don't know if returning `false` is a good idea for a method that usually returns a string. Especially since ~~this method is not always used with that in mind~~ (and this special behaviour is undocumented).

I'd personally rather return `nil` here...

edit: On each use you are checking that there is a URL... darn you modifier form of `if` ^^

floriank replied:
Old habit. I personally dislike some constructs along the lines of "hey, `nil` is falsy". I can change it, though.

NobodysNightmare replied:
You don't need to do that. I am fine with `url.nil?`

My point is that mixed return types ring my statically typed alarm bells. Even more if they are not documented.

NobodysNightmare commented on lib/open_project/text_formatting.rb:
```diff
@@ -379,7 +379,9 @@ def parse_headings(text, _project, _obj, _attr, _only_path, options)
         @parsed_headings << [level, anchor, item]
-        "<a name=\"#{anchor}\"></a>\n<h#{level} #{attrs}>#{content}<a href=\"##{anchor}\" class=\"wiki-anchor\">&para;</a></h#{level}>"
+        url = full_url anchor
+        next "<a name=\"#{anchor}\"></a>\n<h#{level} #{attrs}>#{content}<a href=\"#{url}\" class=\"wiki-anchor\">&para;</a></h#{level}>" if url
```
Except from the (valid) hound warning:

```
next foooooooo if bar?
```

Especially for such long lines this is nasty.

How about

```
if bar?
  next fooooooooo
end
```

I even had to google the meaning of `next 'foo'`, but that is a different topic ^^

NobodysNightmare commented:
> It also removes the links when no request is present (in the case of mailers).

I am missing that in the code. Neither the tests nor the implementation seem to tell me why the hell the request might be missing.

Apparently there is a valid case for that and it should be obvious from the code or its comments (or specs).

I still don't _understand_ it though. Shouldn't it be possible to have a valid link even inside a mail?

I'd either expect `https://example.com/full/path#anchor` to get me out of the mail and into the website **or** simply `#anchor`, because I expect a mail client to be able to handle anchor references inside my mail...

edit: I verified it, my mail client can handle anchors... You received the same mail ^^

floriank commented:
I talked to @ulferts and decided that anchors are not working in emails*.

I'll have it so that the full url is used when a request is present and a simple anchor is used for all other cases, e.g. emails which have been sent through a delayed job.

*people using HTML emails are **wrong**

Commit:
fall back to a normal anchor when no request is present

Signed-off-by: Florian Kraft <f.kraft@finn.de>

NobodysNightmare commented:
> I'll have it so that the full url is used when a request is present and a simple anchor is used for all other cases

Fine by me, this means it will be working in emails ^^

Commit:
repair legacy specs

fix the application helper legacy spec

fix text formatting spec to reflect anchors only mode w/o request

Signed-off-by: Florian Kraft <f.kraft@finn.de>

NobodysNightmare commented:
LGTM :+1: 

Merge as soon as travis agrees...

floriank commented:
@NobodysNightmare merge mich.

myabc commented:
@floriank this solution doesn't work on Rails 4 btw.

```
ActionController::UrlGenerationError:
       No route matches {}
```

31b12a6ead4ef59f7b057aa86280fa64039dc8a9 and ~~9898552e0ff7da54cc6db1f154452211bf0e9130~~ 61d77c721bef8a78a7c52a699e4b00020ffc3673 (PR #2104) should make this PR work on Rails 4. Can you take a look?

floriank commented:
We probably need to find another solution, the `TextFormatting` was not really made for all the angular stuff we are doing.

myabc commented:
@floriank yep.. `TextFormatting` is a pretty crude extraction of code that was formerly in `ApplicationHelper` (d13eba74b9ab334f8567c45707efd5b54b468831). It needs proper refactoring.
